### PR TITLE
Make sure type inference with `a..b` as good as `range(a,b)`

### LINKED ISSUE #21595: Regression in integer fallback
This code compiled fine before 1.0.0-nightly (4be79d6ac 2015-01-23 16:08:14 +0000):

```rust
fn main()
{
	for i in 0..100
	{
		i as f32;
	}
}
```

But with that version, it now gives this error:

```
<anon>:5:3: 5:11 error: the type of this value must be known in this context
<anon>:5 		i as f32;
         		^~~~~~~~
```

### LINKED ISSUE #21649: Type inference for indexing
This example used to work:
```rust
fn main() {
    let array = [true, false];
    
    for i in 0..1 { 
        println!("{}", array[i]);
    }
}
```

Recently you'll get the error: `error: the trait ``core::ops::Index<i32>`` is not implemented for the type ``[bool]`` [E0277]`

### LINKED ISSUE #21672: Type inference with `a..b` is not as good as with `range(a, b)`
### STR

``` rust
fn main() {
    (0..8).map(|i| i == 1u64).collect::<Vec<_>>();
    //~^ error: the type of this value must be known in this context
    (0..8).map(|i| 1u64 == i).collect::<Vec<_>>();
    //~^ error: type mismatch resolving `<core::ops::Range<i32> as core::iter::Iterator>::Item == u64`
    range(0, 8).map(|i| i == 1u64).collect::<Vec<_>>();  // OK
    range(0, 8).map(|i| 1u64 == i).collect::<Vec<_>>();  // OK
}
```

### Version

```
rustc 1.0.0-nightly (458a6a2f6 2015-01-25 21:20:37 +0000)
```

cc @nick29581

### PR DESCRIPTION
**Pull request by edwardw**
The new `::ops::Range` has separated implementations for each of the
numeric types, while the old `::iter::Range` has one for type `Int`.
However, we do not take output bindings into account when selecting
traits. So it confuses `typeck` and makes the new range does not work as
good as the old one when it comes to type inference.

This patch implements `Iterator` for the new range for one type `Int`.
This limitation could be lifted, however, if we ever reconsider the
output types' role in type inference.

Closes #21595
Closes #21649
Closes #21672

### CODE REVIEW

**Comment by emanueLczirai** [comment]
I am curious what happens with a negative range eg. `-1..10`
I'll test this with your PR and let you know. Currently started compiling rust locally.

**Comment by edwardw** [comment]
It should work as expected since after type inference, the numeric fallback will kick in and `-1..10` will then be equivalent to `-1i32..10`.

**Comment by bluss** [comment]
Does this allow `impl Iterator for Range<MyType>` ?  I don't think that's something we require or designed for, but I'm curious nonetheless.

File: src/libcore/iter.rs [hunk 1]
```diff
@@ -2796,94 +2796,58 @@ impl<A: Int> Iterator for RangeStepInclusive<A> {
...
-
-            #[inline]
-            fn next(&mut self) -> Option<$t> {
-                if self.start < self.end {
-                    let result = self.start;
-                    self.start += 1;
-                    return Some(result);
-                }
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<A: Int> ExactSizeIterator for ::ops::Range<A> {}
```
Comment (alexcrichton) [inline src/libcore/iter.rs]: Unfortunately with the definition as-is I don't think that we can have this `ExactSizeIterator` impl here. As defined it's possible for `size_hint` to return `(0, None)` which goes against the contract of `ExactSizeIterator`. 

While we don't have negative bounds, perhaps the `ExactSizeIterator` trait could still just be manually defined for each primitive type?

**Comment by edwardw** [comment]
Hmm, good question. I think so offhand, as long as `Range<T: Int>` and `Range<MyLocalType>` are not considered conflict implementations.

**Comment by emanueLczirai** [comment]
my local compilation reports this:
```
test [compile-fail] compile-fail/xc-private-method2.rs ... ok
test [compile-fail] compile-fail/xcrate-private-by-default.rs ... ok

failures:

---- [compile-fail] compile-fail/range-1.rs stdout ----
	
error: expected error on line 19 not found: `core::iter::Iterator` is not implemented for the type `core::ops::Range<f32>`
status: exit code: 101
command: x86_64-unknown-linux-gnu/stage2/bin/rustc /home/emacs/build/rust/src/test/compile-fail/range-1.rs -L x86_64-unknown-linux-gnu/test/compile-fail --target=x86_64-unknown-linux-gnu -L x86_64-unknown-linux-gnu/test/compile-fail/range-1.stage2-x86_64-unknown-linux-gnulibaux -C prefer-dynamic -o x86_64-unknown-linux-gnu/test/compile-fail/range-1.stage2-x86_64-unknown-linux-gnu --cfg rtopt -O -L x86_64-unknown-linux-gnu/rt
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:15:13: 15:23 error: start and end of range have incompatible types: expected `usize`, found `isize` (expected usize, found isize) [E0308]
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:15     let _ = 0us..10is;
                                                                       ^~~~~~~~~~
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:19:5: 26:8 error: the trait `core::num::Int` is not implemented for the type `f32` [E0277]
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:19     for i in 0f32..42f32 {}
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:20     //~^ ERROR `core::iter::Iterator` is not implemented for the type `core::ops::Range<f32>`
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:21     //~^^ ERROR
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:22     //~^^^ ERROR
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:23     // FIXME(#21528) not fulfilled obligation error should be reported once, not thrice
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:24 
                                                           ...
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:27:17: 27:24 error: the trait `core::marker::Sized` is not implemented for the type `[usize]` [E0277]
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:27     let range = *arr..;
                                                                           ^~~~~~~
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:27:17: 27:24 note: `[usize]` does not have a constant size known at compile-time
/home/emacs/build/rust/src/test/compile-fail/range-1.rs:27     let range = *arr..;
                                                                           ^~~~~~~
error: aborting due to 3 previous errors

------------------------------------------

thread '[compile-fail] compile-fail/range-1.rs' panicked at 'explicit panic', /home/emacs/build/rust/src/compiletest/runtest.rs:1458



failures:
    [compile-fail] compile-fail/range-1.rs

test result: FAILED. 1643 passed; 1 failed; 22 ignored; 0 measured

thread '<main>' panicked at 'Some tests failed', /home/emacs/build/rust/src/compiletest/compiletest.rs:252
/home/emacs/build/rust/mk/tests.mk:729: recipe for target 'tmp/check-stage2-T-x86_64-unknown-linux-gnu-H-x86_64-unknown-linux-gnu-cfail.ok' failed
make: *** [tmp/check-stage2-T-x86_64-unknown-linux-gnu-H-x86_64-unknown-linux-gnu-cfail.ok] Error 101


```

File: src/libcore/iter.rs (hunk 1 above)
Comment (edwardw) [inline src/libcore/iter.rs]: Done, the hybrid implementations for range work!

**Comment by edwardw** [comment]
Thanks @emanueLczirai, the fallout has been fixed.

**Comment by emanueLczirai** [comment]
Compiles successfully for me now. Thanks! (and numeric fallback is as you said earlier: i32)

One question: is that commented FIXME that you removed no longer necessary there? as in, is it fixed or no longer makes sense due to this PR?

**Comment by edwardw** [comment]
It is simply not there any more. I doubt it has anything to do with this PR though.

**Comment by alexcrichton** [comment]
@bors: r+  b9c055c

Thanks!

**Pull request merged**